src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_validators.py

```diff
@@ -83,21 +114,59 @@ def _validate_vm_create_storage_profile(namespace):
+    elif image.lower().endswith('.vhd'):
+        # pylint: disable=redefined-variable-type
+        namespace.storage_profile = StorageProfile.SACustomImage
+    elif not for_scale_set and namespace.managed_os_disk:
+        res = parse_resource_id(namespace.managed_os_disk)
+        name = res['name']
+        rg = res.get('resource_group', namespace.resource_group_name)
+        namespace.managed_os_disk = resource_id(
+            subscription=get_subscription_id(),
+            resource_group=rg,
+            namespace='Microsoft.Compute',
+            type='disks',
+            name=name)
+        namespace.storage_profile = StorageProfile.ManagedSpecializedOSDisk
     else:
         images = load_images_from_aliases_doc()
         matched = next((x for x in images if x['urnAlias'].lower() == image.lower()), None)
-        if matched is None:
-            raise CLIError('Invalid image "{}". Please pick one from {}'.format(
-                image, [x['urnAlias'] for x in images]))
-        namespace.os_publisher = matched['publisher']
-        namespace.os_offer = matched['offer']
-        namespace.os_sku = matched['sku']
-        namespace.os_version = matched['version']
-
-    namespace.os_type = 'windows' if 'windows' in namespace.os_offer.lower() else 'linux'
+        if matched:
+            namespace.os_publisher = matched['publisher']
+            namespace.os_offer = matched['offer']
+            namespace.os_sku = matched['sku']
+            namespace.os_version = matched['version']
+            namespace.storage_profile = (StorageProfile.SAPirImage
+                                         if namespace.use_unmanaged_disk
+                                         else StorageProfile.ManagedPirImage)
+        else:
+            # last try: is it a custom image name?
+            from msrestazure.azure_exceptions import CloudError
+            compute_client = _compute_client_factory()
+            try:
+                compute_client.images.get(namespace.resource_group_name, image)
+                image = namespace.image = _get_ressource_id(image, namespace.resource_group_name,
+                                                            'images', 'Microsoft.Compute')
+                namespace.storage_profile = StorageProfile.ManagedCustomImage
+            except CloudError:
+                err = 'Invalid image "{}". Use a custom image name, id, or pick one from {}'
+                raise CLIError(err.format(image, [x['urnAlias'] for x in images]))
+
+    if namespace.storage_profile == StorageProfile.ManagedCustomImage:
+        res = parse_resource_id(image)
+        compute_client = _compute_client_factory()
+        image_info = compute_client.images.get(res['resource_group'], res['name'])
+        # pylint: disable=no-member
+        namespace.os_type = image_info.storage_profile.os_disk.os_type.value
+        namespace.image_data_disks = image_info.storage_profile.data_disks
 
-    # 2 - Create native disk from PIR image
```

Comment:
I recommend keeping in the comments that show where each of the storage profiles are officially decided.  It will make it easier to maintain I think..